File: tests/journal.lib/firewall_cascade.cc
// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2023 all rights reserved


// get the journal
#include <pyre/journal.h>
// support
#include <cassert>


// type aliases
// chronicler
using chronicler_t = pyre::journal::chronicler_t;
// firewall
using firewall_t = pyre::journal::firewall_t;
// the trash can
using trash_t = pyre::journal::trash_t;


// verify the cascade works correctly for firewalls
int main() {
    // make a channel
    firewall_t parent("test.firewall.parent");
    // its activation state is what's expected
    assert(parent.active());
    // it is fatal
    assert(parent.fatal());
    // and the device is the global default
    assert(parent.device() == chronicler_t::device());
    // turn it off
    parent.deactivate();
    // make it non-fatal
    parent.fatal(false);
    // and set the device to a trash can
    parent.device<trash_t>();

    // make a firewall that's lower in the hierarchy
    firewall_t child("test.firewall.parent.blah.blah.child");
    // make sure its activation state is what's expected
    assert(child.active() == parent.active());
    // it is also non-fatal
    assert(child.fatal() == parent.fatal());
    // and that it inherited the device
    assert(child.device() == parent.device());

    // all done
    return 0;
}


// end of file

File: tests/journal.api/firewall_cascade.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2023 all rights reserved


def test():
    """
    Verify that channels lower in the hierarchy inherit their parent default state
    """
    # get the journal
    import journal

    # make a channel
    parent = journal.firewall(name="test.index.parent")
    # verify that the state is on
    assert parent.active is True
    # that it is fatal
    assert parent.fatal is True
    # and the device is at the default value
    assert parent.device is parent.chronicler.device

    # deactivate it
    parent.active = False
    # and make it non fatal
    parent.fatal = False
    # and set the device to a trash can
    parent.device = journal.trash()

    # lookup a name that is lower in the hierarchy
    child = journal.firewall(name="test.index.parent.blah.blah.child")
    # check that its state is the same as the parent
    assert child.active == parent.active
    assert child.fatal == parent.fatal
    # and that it inherited the device correctly
    assert child.device is parent.device

    # all done
    return


# main
if __name__ == "__main__":
    # run the test
    test()


# end of file
